Python/auth-blocking-functions/main.py

```diff
@@ -0,0 +1,231 @@
+        return identity_fn.BeforeSignInResponse(email_verified=True)
+# [END trustfacebook]
+
+
+def is_suspicious(ip_address):
+    return True
+
+
+# [START ipban]
+@identity_fn.before_user_signed_in()
+def ipban(
+    event: identity_fn.AuthBlockingEvent,
+) -> identity_fn.BeforeSignInResponse | None:
+    if is_suspicious(event.ip_address):
+        raise https_fn.HttpsError(
+            code=https_fn.FunctionsErrorCode.PERMISSION_DENIED,
+            message="IP banned.",
+        )
+# [END ipban]
+
+
+# [START customclaims]
+@identity_fn.before_user_created()
+def setemployeeid(
+    event: identity_fn.AuthBlockingEvent,
+) -> identity_fn.BeforeCreateResponse | None:
+    if (event.credential is not None and
+            event.credential.provider_id == "saml.my-provider-id"):
+        return identity_fn.BeforeCreateResponse(
+            custom_claims={"eid": event.credential.claims["employeeid"]},)
+
+
+@identity_fn.before_user_signed_in()
+def copyclaimstosession(
+    event: identity_fn.AuthBlockingEvent,
+) -> identity_fn.BeforeSignInResponse | None:
+    if (event.credential is not None and
+            event.credential.provider_id == "saml.my-provider-id"):
+        return identity_fn.BeforeSignInResponse(
+            session_claims={
+                "role": event.credential.claims["role"],
+                "groups": event.credential.claims["groups"],
+            })
+# [END customclaims]
+
+
+# [START logip]
+@identity_fn.before_user_signed_in()
+def logip(
+    event: identity_fn.AuthBlockingEvent,
+) -> identity_fn.BeforeSignInResponse | None:
+    return identity_fn.BeforeSignInResponse(
+        session_claims={"signInIpAddress": event.ip_address})
+# [END logip]
+
+
+def analyze_photo_with_ml(url):
+    return 0.42
+
+
+THRESHOLD = 0.7
+PLACEHOLDER_URL = ""
+
+
+# [START sanitizeprofilephoto]
+@identity_fn.before_user_created()
+def sanitizeprofilephoto(
+    event: identity_fn.AuthBlockingEvent,
+) -> identity_fn.BeforeCreateResponse | None:
+    if event.data.photo_url is not None:
+        score = analyze_photo_with_ml(event.data.photo_url)
+        if score > THRESHOLD:
+            return identity_fn.BeforeCreateResponse(photo_url=PLACEHOLDER_URL)
+# [END sanitizeprofilephoto]
+
+
+
+# [START v2CheckForBan]
+# [START v2beforeSignInFunctionTrigger]
+# Block account sign in with any banned account.
+@identity_fn.before_user_signed_in()
+def checkforban(
+    event: identity_fn.AuthBlockingEvent,
+) -> identity_fn.BeforeSignInResponse | None:
+    # [END v2beforeSignInFunctionTrigger]
+    # [START v2readEmailData]
+    # Email passed from the CloudEvent.
+    email = event.data.email if event.data.email is not None else ""
+    # [END v2readEmailData]
+
+    # [START v2documentGet]
+    # Obtain a document in Firestore of the banned email address.
+    firestore_client: google.cloud.firestore.Client = firestore.client()
+    doc = firestore_client.collection("banned").document(email).get()
```

Comment:
Can we use the Firebase Admin SDK here instead?

Reply:
We do, that's just for typing/Intellisense